Improve documentation relating to `Frustum` and `HalfSpace`

Selene-Amanita opened the pull request:
# Objective

This PR's first aim is to fix a mistake in `HalfSpace`'s documentation.

When defining a `Frustum` myself in bevy_basic_portals, I realised that the "distance" of the `HalfSpace` is not, as the current doc defines, the "distance from the origin along the normal", but actually the opposite of that.

See the example I gave in this PR.

This means one of two things:
1. The documentation about `HalfSpace` is wrong (it is either way because of the `n.p + d > 0` formula given later anyway, which is how it behaves, but in that formula `d` is indeed the opposite of the "distance from the origin along the normal", otherwise it should be `n.p > d`)
2. The distance is supposed to be the "distance from the origin along the normal" but when used in a Frustum it's used as the opposite, and it is a mistake
3. Same as 2, but it is somehow intended

Since I think `HalfSpace` is only used for `Frustum`, and it's easier to fix documentation than code, I assumed for this PR we're in case number 1. If we're in case number 3,  the documentation of `Frustum` needs to change, and in case number 2, the code needs to be fixed.

While I was at it, I also :
- Tried to improve the documentation for `Frustum`, `Aabb`, and `VisibilitySystems`, among others, since they're all related to `Frustum`.
- Fixed documentation about frustum culling not applying to 2d objects, which is not true since https://github.com/bevyengine/bevy/pull/7885

## Remarks and questions

- What about a `HalfSpace` with an infinite distance, is it allowed and does it represents the whole space? If so it should probably be mentioned.
- I referenced the `update_frusta` system in `bevy_render::view::visibility` directly instead of referencing its system set, should I reference the system set instead? It's a bit annoying since it's in 3 sets.
- `visibility_propagate` is not public for some reason, I think it probably should be, but for now I only documented its system set, should I make it public? I don't think that would count as a breaking change?
- Why is `Aabb` inserted by a system, with `NoFrustumCulling` as an opt-out, instead of having it inserted by default in `PbrBundle` for example and then the system calculating it when it's added? Is it because there is still no way to have an optional component inside a bundle?

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
 /// An axis-aligned bounding box.
+///
+/// It represents a box covering the local space occupied by the entity, with faces
+/// orthogonal to the axis. It is used as a component on an entity to determine
```
I can add "local" in front of "axis" on the second line too, would that be enough? Maybe add a parenthesis after the first "local" for "(relative to the entity)"?

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,11 +93,29 @@ impl Sphere {
+/// Each instance of this type is characterized by:
+/// - the bisecting plane's unit normal, normalized and pointing "inside" the half-space,
+/// - the distance from the bisecting plane to the origin of 3D space along the normal,
+///
+/// The distance can also be seen as:
+/// - the distance from the origin of 3D space to the bisecting plane along the inverse of the normal,
+/// - the opposite of the distance from the origin of 3D space to the bisecting plane along the normal.
```
To me, it would make more sense intuitively that the distance would be from the origin to the plane along the normal (so the opposite of what it is) so I wanted to put the emphasis on that.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,11 +93,29 @@ impl Sphere {
+/// Any point `p` is considered to be within the `HalfSpace` when the length of the projection
+/// of p on the normal is greater or equal than the opposite of the distance,
+/// meaning: if the equation `n.p + d >= 0` is satisfied,
+/// where `n` is the normal, `d` the distance, and `n.p` is a dot product.
```
Changed to `normal.dot(p) + distance >= 0.` in https://github.com/bevyengine/bevy/pull/9136/commits/9714243737b741d2abe02b8d7a60a2216eb82bb9

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -109,8 +139,9 @@ impl HalfSpace {
-    /// This distance helps determine the position of a point `p` on the bisecting plane, as per the equation `n.p + d = 0`.
+    /// Returns the distance from the bisecting plane to the origin along the plane's unit normal vector.
+    /// This distance helps determine the position of a point `p` relative to the
+    /// bisecting plane, as per the equation `n.p + d >= 0`.
```
Fixed in https://github.com/bevyengine/bevy/pull/9136/commits/9714243737b741d2abe02b8d7a60a2216eb82bb9 the same way as https://github.com/bevyengine/bevy/pull/9136#discussion_r1264738811

Selene-Amanita commented:
Thanks for reviewing!

> Why are UpdateOrthographicFrusta, UpdatePerspectiveFrusta and UpdateProjectionFrusta separate system sets?

I... don't know actually, they're used for ordering but I think this could be made with the systems themselves inside the same SystemSet?

I'll make another PR to make `visibility_propagate` public later (except if someone else wants to do it)

nicopap commented:
You'll get my review tomorrow.

mockersf commented on crates/bevy_render/src/camera/projection.rs:
```diff
@@ -53,6 +53,9 @@ impl<T: CameraProjection + Component + GetTypeRegistration> Plugin for CameraPro
 /// to recompute the camera projection matrix of the [`Camera`] component attached to
 /// the same entity as the component implementing this trait.
 ///
+/// Components implementing this trait include: [`Projection`], [`PerspectiveProjection`]
+/// and [`OrthographicProjection`].
+///
```
this is hard to maintain doc, and already done by docs.rs, see https://docs.rs/bevy/latest/bevy/render/camera/trait.CameraProjection.html#implementors

Selene-Amanita replied:
Removed in https://github.com/bevyengine/bevy/pull/9136/commits/dee7a95f4e34e2534c35ebcbf8a292c764dd5519

mockersf commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
 
 /// An axis-aligned bounding box.
+///
+/// It represents a box covering the local space occupied by the entity, with faces
+/// orthogonal to the local axis. It is used as a component on an entity to determine
+/// if it should be rendered by a [`Camera`](crate::camera::Camera) entity if it
+/// intersects with its [`Frustum`], in a process called frustum culling.
```
saying it's used for frustum culling is very limiting to `Aabb`, which is why it's in the primitive module and not in view. `Aabb` are often used for collisions, for debug, for sorting in space, ...

Selene-Amanita replied:
Is it really? Because it's not inserted to entities with `NoFrustumCulling`, so if it is, shouldn't `NoFrustumCulling` be handled differently?

mockersf replied:
In Bevy it's used just for frustum culling, but an aabb is a general type that have many uses. It's a component that users can use however they want, we shouldn't restrict it to frustum culling in its doc

Selene-Amanita replied:
I agree, but since it's linked to `NoFrustumCulling`, I don't see it being used for something else.

Selene-Amanita replied:
Addressed partially by https://github.com/bevyengine/bevy/pull/9136/commits/936a9122b44575c948e59c443c37a13bbc9d01c6

mockersf commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,11 +93,28 @@ impl Sphere {
+/// It includes all the points from the bisecting plane towards `NEG_Z`, and the distance
+/// from the plane to the origin is `-8.0` along `NEG_Z`.
+///
+/// It is used to define a [`Frustum`].
```
among other things

Selene-Amanita replied:
Is it used for something else in Bevy? I didn't want to imply it's the only usage, I just wanted to say it's only used for that in Bevy.

Selene-Amanita replied:
Changed to `It is used to define a [`Frustum`], and for light computation.` in https://github.com/bevyengine/bevy/pull/9136/commits/dee7a95f4e34e2534c35ebcbf8a292c764dd5519

Is that okay?

mockersf replied:
Same as aabb, those are general types and component, doc shouldn't restrict how they are used.

I'm not sure I see the point of documenting that here anyway. being able to go from the frustrum doc to the half space is interesting, not the reverse in my opinion

Selene-Amanita replied:
I think it's good to have a concrete example of what it is used for instead of just having it and not knowing how it can be used. I don't really see how my sentence implies it can't be used for something else?

nicopap reviewed (commented):
I think the phrasing could be streamlined.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
+/// if it should be rendered by a [`Camera`](crate::camera::Camera) entity if it
+/// intersects with its [`Frustum`], in a process called frustum culling.
+///
+/// It will be added automatically to entities that could be subject to frustum
+/// culling, and don't have the [`NoFrustumCulling`](crate::view::visibility::NoFrustumCulling)
```
It isn't but I'm not sure myself why would you want to opt out of it actually.

nicopap replied:
See #4971

Selene-Amanita replied:
This is because in animation the Mesh's vertices positions are updated, but the Aabb is not, right?

Lighting/reflection is for when you have a Mesh with a reflective material and you want to see something in it that is out of frustum, right?

I'm trying to write meaningful examples.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
+/// component, using the systems in
+/// [`CalculateBounds`](crate::view::visibility::VisibilitySystems::CalculateBounds).
+///
+/// It won't be updated automatically if the space occupied by the entity changes.
```
I can add that in a "for example" section after (I think it's a good idea), but Aabb don't apply only to `Handle<Mesh>`.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
 use bevy_utils::HashMap;
 
 /// An axis-aligned bounding box.
```
It isn't applied only to `Handle<Mesh>`, people can find what it applies to if they follow the doc of `CalculateBounds`, but maybe I can add a "for example" section?

Selene-Amanita replied:
Mentionned it here:

```rust
/// It will be added automatically by the systems in [`CalculateBounds`] to entities that:
/// - could be subject to frustum culling, for example [`Handle<Mesh>`] or `Sprite`,
/// - don't have the [`NoFrustumCulling`] component.
```

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,11 +93,28 @@ impl Sphere {
+///
+/// Each instance of this type is characterized by:
+/// - the bisecting plane's unit normal, normalized and pointing "inside" the half-space,
+/// - the signed distance from the bisecting plane to the origin of 3D space along the normal.
```
"along the normal" applies to the distance, not the origin of 3D space

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,11 +93,28 @@ impl Sphere {
+/// - the opposite of the distance from the origin of 3D space to the bisecting plane along the normal.
+///
+/// Any point `p` is considered to be within the `HalfSpace` when the length of the projection
+/// of p on the normal is greater or equal than the opposite of the distance,
```
The plane doesn't go through the origin? I don't understand sorry.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -123,9 +154,24 @@ impl HalfSpace {
-/// Half spaces are ordered left, right, top, bottom, near, far.
-/// The normal vectors of the half spaces point towards the interior of the frustum.
+/// A frustum defined as the intersection of 6 [`HalfSpace`].
+/// Usually a six sided polyhedron with sides inside the bisecting planes of the half-spaces.
```
It can be a cuboid for orthogonal projection, or I guess technically it could have less than six sides or extend to infinity.

But I did say "Usually" and a Cuboid is a special case of truncated pyramid I guess.

According to wikipedia, the real name of the truncated pyramid is either "Quadrilateral frustum" (which doesn't help much here) or "apex-truncated square pyramid" which I think works? Are you okay with the last one?

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -4,6 +4,18 @@ use bevy_reflect::Reflect;
+///
+/// It represents a box covering the local space occupied by the entity, with faces
+/// orthogonal to the axis. It is used as a component on an entity to determine
+/// if it should be rendered by a [`Camera`](crate::camera::Camera) entity if it
+/// intersects with its [`Frustum`], in a process called frustum culling.
```
`It is used as a component on an entity during "frustum culling",  a process to determine if the entity should be rendered [...]`?

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -109,8 +138,10 @@ impl HalfSpace {
-    /// This distance helps determine the position of a point `p` on the bisecting plane, as per the equation `n.p + d = 0`.
+    /// Returns the signed distance from the bisecting plane to the origin along
+    /// the plane's unit normal vector.
+    /// This distance helps determine the position of a point `p` relative to the
```
Honestly I would kinda be in favor of removing that sentence completely, it's already explained in the `HalfSpace` doc and you can do nothing with the distance alone anyway you need the normal for the formula, if this formula should be copied somewhere I would argue it should be in `normal_d`.

This sentence was written for `Plane` to know on which side of the plane we are.

Selene-Amanita commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -123,9 +154,24 @@ impl HalfSpace {
+/// the [`NoFrustumCulling`](crate::view::visibility::NoFrustumCulling) component.
+///
+/// The frustum component is typically added by adding a bundle, either the `Camera2dBundle`
+/// or the `Camera3dBundle`, and is typically updated automatically by
+/// [`update_frusta`](crate::view::visibility::update_frusta) from the
+/// [`CameraProjection`](crate::camera::CameraProjection) component of the camera entity.
```
My French brain struggles to make short sentences :smile: (I agree)

Selene-Amanita commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -267,6 +277,13 @@ pub fn calculate_bounds(
     }
 }
 
+/// System updating the [`Frustum`] component of an entity, typically a [`Camera`],
+/// using its [`GlobalTransform`] and the projection matrix from its [`CameraProjection`]
+/// component.
```
Maybe I should still mention `Camera`, or `Camera3dBundle`/`Camera2dBundle`, to add context?

I guess the context is in the doc of `Frustum` and `CameraProjection` already.

Selene-Amanita commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -253,6 +259,10 @@ impl Plugin for VisibilityPlugin {
     }
 }
 
+/// System calculating and inserting an [`Aabb`] component to entities with a
```
Changed to `System to calculate and insert` to keep the information that it's a system (same for the other ones).

Selene-Amanita replied:
Changed the sentence bellow to keep the information that it's a system: `Used in system set` -> `This system is used [...]`

Selene-Amanita commented:
@nicopap I addressed most of the stuff in https://github.com/bevyengine/bevy/pull/9136/commits/19c3c8cdf568dd2c68fe794d16276378fe754307 and put an edit to my comments to your reviews for how I handled it. If you're okay with the changes you can put a :+1: and I'll resolve the conversation.

Selene-Amanita commented:
> Is HalfSpace supposed to be an open or closed set, meaning: should the bisecting plane be in it or out of it? From my tests (and reading of intersects_obb and intersects_sphere), I think it is supposed to be in, hence why I changed the documentation to clarify that and changed the formula to n.p + d >= 0, but that could be because of rounding errors, and I think it doesn't really matter anyway, but I wanted to clarify.

After testing and analysis of the `Frustum` code (which excludes stuff with `n.p + d <= 0`, so includes `>0`), HalfSpace seems to be an open set instead (at least from the perspective of `Frustum`, there is no actual method in `HalfSpace to say if a vertex is in or not). Fixed in https://github.com/bevyengine/bevy/pull/9136/commits/aeaef2d8cd389e14066f45f2f6dd6a3ca6f8337b .

alice-i-cecile commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -81,20 +106,37 @@ impl Sphere {
+/// It includes all the points from the bisecting plane towards `NEG_Z`, and the distance
+/// from the plane to the origin is `-8.0` along `NEG_Z`.
+///
+/// It is used to define a [`Frustum`], and for light computation, for example.
```
```suggestion
/// It is used to define a [`Frustum`], but is also a useful mathematical primitive for rendering tasks such as  light computation.
```

alice-i-cecile commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -109,23 +151,46 @@ impl HalfSpace {
-/// A frustum made up of the 6 defining half spaces.
-/// Half spaces are ordered left, right, top, bottom, near, far.
-/// The normal vectors of the half spaces point towards the interior of the frustum.
+/// A frustum defined as the intersection of 6 [`HalfSpace`].
```
```suggestion
/// A region of 3D space defined by the intersection of 6 [`HalfSpace`].
```

alice-i-cecile commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -109,23 +151,46 @@ impl HalfSpace {
-/// The normal vectors of the half spaces point towards the interior of the frustum.
+/// A frustum defined as the intersection of 6 [`HalfSpace`].
+///
+/// Usually an apex-truncated square pyramid (a pyramid without the top) or a cuboid.
```
```suggestion
/// Frustums are typically an apex-truncated square pyramid (a pyramid without the top) or a cuboid.
```

alice-i-cecile commented on crates/bevy_render/src/primitives/mod.rs:
```diff
@@ -109,23 +151,46 @@ impl HalfSpace {
+///
+/// A frustum component is used on an entity with a [`Camera`] component
+/// to determine which entities will be considered for rendering by this camera.
+/// All entities with an [`Aabb`] component that does not intersect with the frustum
```
```suggestion
/// All entities with an [`Aabb`] component that are not contained by (or overlapping the boundary of) the frustum
```

Selene-Amanita replied:
I changed "overlapping" to "crossing" since technically the boundary is not part of the frustum and if the Aabb just touches the boundary they overlap but the entity will not be considered.

alice-i-cecile reviewed (changes requested):
Needs a bit more work, but definitely a marked improvement.
